### PR TITLE
Assignment fifteen IDispose pattern

### PR DESCRIPTION
**Pull request by Sruthi-Subaraja**
ASSIGNMENT 15 - IDISPOSE PATTERN
TASK 1 - Understanding the difference between managed and unmanaged resource.
TASK 2 - Implementing the IDisposable interface.
TASK 3 - Implementing the dispose pattern.
TASK 4 - Using finalizers.

### CODE REVIEW

File: src/IDisposePattern/Program.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            var managed = new ManagedResource();
+            managed.AddData("Hello, world!");
+            Console.WriteLine(managed.Data[0]);
+
+            // managed.Dispose();
+            // managed.ClearData();
+            Console.WriteLine(managed.Data.Count);
+            string currentDirectory = Directory.GetCurrentDirectory();
+            string filePath = Path.Combine(currentDirectory, "test.txt");
```
Comment (Tharun-Sundar) [inline src/IDisposePattern/Program.cs:26]: Where is filePath used?

File: tests/unit/IDisposePattern/UnmanagedResourceClassTest/UnmanagedResourceClassTest.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        [Fact]
+        public void EmptyFile_WritingToFile_ContentIsWrittenToFile()
+        {
+            UnmanagedResource unmanagedResource = new UnmanagedResource("test.txt");
+            string expectedValue = "Hello";
+
+            unmanagedResource.WriteToFile("Hello");
+            unmanagedResource.Writer.Close();
+            StreamReader streamReader = new StreamReader("test.txt");
```
Comment (Tharun-Sundar) [inline tests/unit/IDisposePattern/UnmanagedResourceClassTest/UnmanagedResourceClassTest.cs:15]: StreamReader should not be used here. It should be mocked

**Review by Tharun-Sundar — CHANGES_REQUESTED** [review]

File: tests/unit/IDisposePattern/ManagedResourcesClassTest/ManagedResourceClassTest.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        [Fact]
+        public void EmptyList_AddingDataToList_ListHasData()
+        {
+            ManagedResource managedResource = new ManagedResource();
+            int expectedValue = 1;
+
+            managedResource.AddData("Hello");
+            
+            Assert.Equal(expectedValue, managedResource.Data.Count());
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply tests/unit/IDisposePattern/ManagedResourcesClassTest/ManagedResourceClassTest.cs:15]: Okay

File: src/IDisposePattern/Program.cs (hunk 1 above)
Comment (Sruthi-Subaraja) [reply src/IDisposePattern/Program.cs:26]: It is not needed, I removed it.

File: src/IDisposePattern/ManagedResource.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        /// <summary>
+        /// Finalizes an instance of the <see cref="ManagedResource"/> class.
+        /// </summary>
+        ~ManagedResource() => this.Dispose(false);
+
+        /// <summary>
+        /// Gets or sets a list of string.
+        /// </summary>
+        /// <value>A list of string.</value>
+        public List<string> Data { get; set; } = new List<string>();
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/IDisposePattern/ManagedResource.cs:23]: I'll need it outside of the class to test it and print it's values, so I updated it to private set now.

File: tests/unit/IDisposePattern/UnmanagedResourceClassTest/UnmanagedResourceClassTest.cs (hunk 2 above)
Comment (Sruthi-Subaraja) [reply tests/unit/IDisposePattern/UnmanagedResourceClassTest/UnmanagedResourceClassTest.cs]: Okay done

File: tests/unit/IDisposePattern/ManagedResourcesClassTest/ManagedResourceClassTest.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace ManagedResourcesClassTest
+{
+    public class ManagedResourceClassTest
+    {
+        [Fact]
+        public void EmptyList_AddingDataToList_ListHasData()
+        {
+            ManagedResource managedResource = new ManagedResource();
+            int expectedValue = 1;
```
Comment (Tharun-Sundar) [inline tests/unit/IDisposePattern/ManagedResourcesClassTest/ManagedResourceClassTest.cs:11]: Variable name can be `expectedCount`

**Review by Tharun-Sundar — APPROVED** [review]

File: tests/unit/IDisposePattern/ManagedResourcesClassTest/ManagedResourceClassTest.cs (hunk 5 above)
Comment (Sruthi-Subaraja) [reply tests/unit/IDisposePattern/ManagedResourcesClassTest/ManagedResourceClassTest.cs]: Okay, changes made.

**Pull request merged**